yoyodyne/models/modules/lstm.py

```diff
@@ -226,25 +232,29 @@ def name(self) -> str:
 
 
 class HardAttentionLSTMDecoder(LSTMDecoder):
+    """Zeroth-order HMM hard attention LSTM decoder."""
+
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         # Activates emission probs.
         self.output_proj = nn.Sequential(
             nn.Linear(self.output_size, self.output_size), nn.Tanh()
         )
-        # Project transition probabilities to depth of module.
+        # Projects transition probabilities to depth of module.
         self.scale_encoded = nn.Linear(
             self.decoder_input_size, self.hidden_size
         )
 
+    # TODO: needs docs.
```

Comment:
Note for myself.

Reply:
def _alignment_step(
        self,
        decoded: torch.Tensor,
        encoder_out: torch.Tensor,
        encoder_mask: torch.Tensor,
    ) -> torch.Tensor:
        """Creates alignment matrix for current time-step.

        Given current encoder representation and the decoder
            representations at current time step, calculates alignment
            scores between all potential source sequence pairings.
            These alignments are used to predict likelihood of state
            transition for current decoder output.
            
             See: https://aclanthology.org/P19-1148/.

        Args:
            decoded (torch.Tensor): Output from decoder for
                current time step of shape B x 1 x decoder_dim
            encoder_out (torch.Tensor): encoded input sequence of shape
                B x seq_len x encoder_dim.
            encoder_mask (torch.Tensor): mask for the encoded input batch of
                shape B x seq_len.
        
                
        Returns:
            torch.Tensor of alignment scores across source sequence
                of shape B x seq_len x seq_len.
        """

        alignment_scores = torch.bmm(
            self.scale_encoded(encoder_out), decoded.transpose(1, 2)
        ).squeeze(-1)
        # Gets probability of alignments.
        alignment_probs = nn.functional.softmax(alignment_scores, dim=-1)
        # Mask padding.
        alignment_probs = alignment_probs * (~encoder_mask) + 1e-7
        alignment_probs = alignment_probs / alignment_probs.sum(
            dim=-1, keepdim=True
        )
        # Expands over all time steps. Log probs for quicker computation.
        return (
            alignment_probs.log()
            .unsqueeze(1)
            .expand(-1, encoder_out.shape[1], -1)
        )